subsys/bluetooth/host/hci_core.c

```diff
@@ -1744,6 +1719,31 @@ int bt_unpair(u8_t id, const bt_addr_le_t *addr)
 	return 0;
 }
 
+int bt_unpair(u8_t id, const bt_addr_le_t *addr)
+{
+	struct bt_conn *conn;
+	int err, i;
+
+	if (id >= CONFIG_BT_ID_MAX) {
+		return -EINVAL;
+	}
+
+	if (!addr || !bt_addr_le_cmp(addr, BT_ADDR_LE_ANY)) {
+		for (i = 0; i < CONFIG_BT_MAX_CONN; i++) {
+			conn = bt_conn_lookup_id(i);
```

Comment:
I'm not following the logic here. Removing all pairing information means that we need to remove the information regardless of there being a connection or not, yet here you're looking up a connection object before you remove anything?

Reply:
Right. Is there anything else we can use to iterate over?

Reply:
It depends on the module. E.g. in the case of bt_keys_clear_all(), well you can see what that did. Since the root of this all cleanup is an API called "unpair" maybe it's fair to pick up (and act upon) the remote addresses during the iterating through all available keys?

Reply:
I tried using `bt_keys_foreach` but the returned `struct bt_keys` always had its `addr` member set to 0. `bt_foreach_bond` works, but is that only for bonded peers? I can't see any bond-specific stuff in it.

Reply:
Yes, sorry. `bt_keys_foreach` iterates the `keys.c`-internal array "blindly" i.e. doesn't care if it contains valid data. `bt_foreach_bond()` on the other hand (and I apologise for the inconsistent naming) does consider whether the entry in the array has valid data or not.

Reply:
>  I can't see any bond-specific stuff in it.

It's kind of implicit. For pairings using "no bonding" flags we don't populate the bt_keys struct, i.e. for any entries `bt_foreach_bond()` selects we know that the device is bonded.

Reply:
I just pushed a version using `bt_foreach_bond()`, but that won't work either (for the paired-but-not-bonded case)?